lib/operations/add_user.js

```diff
@@ -1,184 +1,173 @@
+    let command = {
+      createUser: username,
+      customData: options.customData || {},
+      roles: roles,
+      digestPassword
+    };
+
+    // No password
+    if (typeof password === 'string') {
+      command.pwd = userPassword;
+    }
 
-  // If not roles defined print deprecated message
-  if (roles.length === 0) {
-    console.log('Creating a user without roles is deprecated in MongoDB >= 2.6');
-  }
+    super(db, command, commandOptions);
 
-  // Get the error options
-  const commandOptions = { writeCommand: true };
-  if (options['dbName']) commandOptions.dbName = options['dbName'];
-
-  // Add maxTimeMS to options if set
-  if (maxTimeMS != null) commandOptions.maxTimeMS = maxTimeMS;
-
-  // Check the db name and add roles if needed
-  if (
-    (db.databaseName.toLowerCase() === 'admin' || options.dbName === 'admin') &&
-    !Array.isArray(options.roles)
-  ) {
-    roles = ['root'];
-  } else if (!Array.isArray(options.roles)) {
-    roles = ['dbOwner'];
+    this.username = username;
+    this.password = password;
+    this.finalOptions = Object.assign({}, options);
   }
 
-  const digestPassword = db.s.topology.lastIsMaster().maxWireVersion >= 7;
+  execute(callback) {
+    const db = this.db;
+    const username = this.username;
+    const password = this.password;
+    let finalOptions = this.finalOptions;
 
-  // Build the command to execute
-  let command = {
-    createUser: username,
-    customData: customData,
-    roles: roles,
-    digestPassword
-  };
+    let Db = loadDb();
 
-  // Apply write concern to command
-  command = applyWriteConcern(command, { db }, options);
+    // Error out if digestPassword set
+    if (finalOptions.digestPassword != null) {
+      return callback(
+        toError(
+          "The digestPassword option is not supported via add_user. Please use db.command('createUser', ...) instead for this option."
+        )
+      );
+    }
 
-  let userPassword = password;
+    // Attempt to execute auth command
+    super.execute((err, r) => {
+      if (!err) {
+        return handleCallback(callback, err, r);
+      }
+
+      if (err && err.ok === 0 && err.code === undefined) {
+        // We need to perform the backward compatible insert operation
+        finalOptions = applyWriteConcern(Object.assign({}, finalOptions), { db }, finalOptions);
+
+        // Use node md5 generator
+        const md5 = crypto.createHash('md5');
+        // Generate keys used for authentication
+        md5.update(username + ':mongo:' + password);
+        const userPassword = md5.digest('hex');
+
+        // If we have another db set
+        const dbToUse = finalOptions.dbName
+          ? new Db(finalOptions.dbName, db.s.topology, db.s.options)
+          : db;
+
+        // Fetch a user collection
+        const collection = dbToUse.collection(CONSTANTS.SYSTEM_USER_COLLECTION);
+
+        // Check if we are inserting the first user
+        count(collection, {}, finalOptions, (err, count) => {
```

Comment:
I am a little worried about the `count`, `collection.find`, and `updateOne` calls here. Hypothetically, these are all separate operations, right? And in the case of the `collection.find`, we are incurring an extra `executeOperation`, right?

I think it is fine for now, but I do wonder if eventually these will need to be converted to distinct sub-operations.

Reply:
You're definitely right. It looks like the [Java driver](https://github.com/mongodb/mongo-java-driver/blob/master/driver-core/src/main/com/mongodb/operation/CreateUserOperation.java#L107) doesn't have all of these operations inside of `createUser`. It appears that we only have `count`, `collection.find`, and `updateOne` here because "We need to perform the backward compatible insert operation." Where did that come from?